sherpa/astro/xspec/tests/test_xspec.py

```diff
@@ -50,60 +166,306 @@ def test_create_model_instances(self):
-        models.remove('XSAdditiveModel')
-        models.remove('XSTableModel')
-
-        xx = numpy.arange(0.1, 11.01, 0.01, dtype=float)
-        xlo = numpy.array(xx[:-1])
-        xhi = numpy.array(xx[1:])
+        models = get_xspec_models(xs)
+
+        egrid, elo, ehi, wgrid, wlo, whi = make_grid()
         for model in models:
             cls = getattr(xs, model)
-            foo = cls(model)  # use an identifier in case there is an error
-            # The model checks that the values are all finite.
-            vals = foo(xlo,xhi)
+            mdl = cls(model)  # use an identifier in case there is an error
+
+            # The model checks that the values are all finite,
+            # so there is no need to check that the output of
+            # mdl does not contain non-finite values.
+            #
+            evals1 = mdl(egrid)
+            evals2 = mdl(elo, ehi)
+
+            wvals1 = mdl(wgrid)
+            wvals2 = mdl(wlo, whi)
+
+            emsg = "{} model evaluation failed: ".format(model)
+
+            # It might be expected that the test should be
+            #   assert_allclose(evals1[:-1], evals2)
+            # to ensure there's no floating-point issues,
+            # but in this case the grid and parameter
+            # values *should* be exactly the same, so the
+            # results *should* be exactly equal, hence
+            # the use of assert_array_equal
+            assert_array_equal(evals1[:-1], evals2,
+                               err_msg=emsg + "energy comparison")
+            assert_array_equal(wvals1[:-1], wvals2,
+                               err_msg=emsg + "wavelength comparison")
+
+            # When comparing wavelength to energy values, have
+            # to use allclose since the bins are not identically
+            # equal.
+            assert_allclose(evals1, wvals1,
+                            err_msg=emsg + "energy to wavelength")
+
+    def test_xspec_models_noncontiguous2(self):
+        """Note that there is no test that the non-contiguous grid
+        results are similar to the result from using a contiguous
+        grid and then filtering out the missing bins. The way that
+        some models are implemented make this a tricky test to
+        write (due to numerical tolerances), as bins at the
+        edges may not match well.
+        """
+
+        import sherpa.astro.xspec as xs
+        models = get_xspec_models(xs)
+
+        elo, ehi, wlo, whi = make_grid_noncontig2()
+        for model in models:
+            cls = getattr(xs, model)
+            mdl = cls(model)  # use an identifier in case there is an error
+
+            evals2 = mdl(elo, ehi)
+            wvals2 = mdl(wlo, whi)
+            emsg = "{} non-contiguous model evaluation failed: ".format(model)
+
+            assert_allclose(evals2, wvals2,
+                            err_msg=emsg + "energy to wavelength")
+
+    def test_tablemodel_checks_input_length(self):
```

Comment:
This is missing the decorator

    @unittest.skipIf(test_data_missing(), "required test data missing")